protoc-gen-kotlin/protoc-gen-kotlin-common/src/main/kotlin/pbandk/gen/CodeGenerator.kt

```diff
@@ -27,15 +27,19 @@ open class CodeGenerator(val file: File, val kotlinTypeMappings: Map<String, Str
     }
 
     protected fun writeEnumType(type: File.Type.Enum) {
-        // Enums are data classes w/ a single value and a companion object with known values
-        line().line("data class ${type.kotlinTypeName}(override val value: Int) : pbandk.Message.Enum {").indented {
-            line("companion object : pbandk.Message.Enum.Companion<${type.kotlinTypeName}> {").indented {
-                type.values.forEach { line("val ${it.kotlinValueName} = ${type.kotlinTypeName}(${it.number})") }
-                line()
-                line("override fun fromValue(value: Int) = when (value) {").indented {
-                    type.values.forEach { line("${it.number} -> ${it.kotlinValueName}") }
-                    line("else -> ${type.kotlinTypeName}(value)")
-                }.line("}")
+        // Enums are data classes w/ a value and a name, and a companion object with known values
+        line().line("sealed class ${type.kotlinTypeName}(override val value: Int, override val name: String) : pbandk.Message.NamedEnum {").indented {
+            line("override fun equals(other: kotlin.Any?) = other is ${type.kotlinTypeName} && other.value == value")
+            line("override fun hashCode() = value.hashCode()")
+            line("override fun toString() = \"${type.kotlinTypeName}.\$name(value=\$value)\"")
+            line()
+            type.values.forEach { line("object ${it.kotlinValueClassName} : ${type.kotlinTypeName}(${it.number}, \"${it.kotlinValueStringName}\")") }
+            line("class Unrecognized(value: Int) : ${type.kotlinTypeName}(value, \"UNRECOGNIZED\")")
```

Comment:
🔧 I'm willing to be swayed the other way on this, but my inclination is to make the `name` property a `String?` rather than `String` and represent `Unrecognized` enum values with a `null` `name`. Having it as a `String?` will force the code that uses this enum to deal with the fact that it could've received an unknown enum value on the wire and hence it doesn't know the enum value's name, only its number. That seems in the spirit of the protobuf ethos of having an API that supports forward-compatibility. Whereas your current approach requires "correct" code to remember to check whether it's dealing with an `Unrecognized` enum value before handling the `name` property.

Reply:
Sure, I can change it to be `String?`, and we can see the fallout.  I imagine if we are inserting into the database, we might do something like `type.name ?: "UNKNOWN"` anyway, but maybe we'd just throw a `BadRequestException` before then anyway... either way is fine, so will change.

Reply:
The database case is actually a great example (that didn't occur to me) for why to make this nullable. We want the compiler to force us to choose what to do with unknown name fields rather than blindly inserting `UNKNOWN` into the database (which we would be unable to map back out of the database later since we wouldn't know the numeric value to associate with the `UNKNOWN` name).